### PR TITLE
tests: Add fuzz testing for libfuzzer and oss fuzz

### PR DESCRIPTION
**Pull request by pdxjohnny**
* Removed struct TSS2_SYS_CONTEXT definition from
  src/tss2-sys/sysapi_util.h as it caused errors with C++ build.
* travis will build fuzz tests to make sure they don't bit rot if
  building with clang.
* Dockerfile uses targets and can build just fuzz tests.
* Added Makefile-fuzz.am and generating Makefile-fuzz-generated.am
  which build fuzz tests.
* Added python script gen_fuzz.py which reads include/tss2/tss2_sys.h
  and generates a fuzz target for all Prepare and Complete calls.
* Added test/fuzz/main-sapi.cpp which defines a libfuzzer target used to
  fuzz SAPI calls.
* Added a TCTI for fuzzing which stores fuzzer generated data which
  populates structures used for SAPI Prepare calls or feed to Complete
  and subsequently unmarshalled.
* Modified integration sapi context util to enable static linking
  against a TCTI.
* Modified integration test options to add the fuzzing TCTI type.

- [x] See about MAINTAINERS path required review with github. https://blog.github.com/2017-07-06-introducing-code-owners/
  - Serarate change. SHould probably be in separate PR
- [x] Investigate ifdefs in `sapi-context-util.c` and why John has decided
  to disable other tctis in order to build the fuzzing tcti.
  - On link `multiple definition of 'tss2_tcti_info'` error when more than one tcti is enabled
    and `--disable-shared` is given to configure (+@flihp)
- [x] Review where comments need to be added.
- [x] Do not generate fuzz tests by default in `bootstrap`
- [x] `struct TSS2_SYS_CONTEXT;` remove from `sysapi_util.h` in seperate PR (#1244)
- [x] Split into separate commits

### CODE REVIEW

**Comment by pdxjohnny** [comment]
Ah yes, I have that enabled and forgot about it. I'm also looking for feedback like you gave about not having things outside of `/test/fuzzing`,  thanks.

**Comment by pdxjohnny** [comment]
In order to feed fuzzing data into SAPI layer I opted for creating a TCTI

Right now this seems to be the only reasonable way of passing data from the fuzzer function to the `_Complete` and `_Prepare` functions. I'm not sure how to take that out of `src/` though or if we can just leave it there.

**Comment by pdxjohnny** [comment]
`script/gen_fuzz.py` generates files which fuzz SAPI `_Prepare` and `_Complete` calls as follows.

> pinging @flihp

### Sample `_Complete`

Fuzzing TCTI feeds data up to `_Complete` functions

```c
/* SPDX-License-Identifier: BSD-2 */
/***********************************************************************
 * Copyright (c) 2018, Intel Corporation
 *
 * All rights reserved.
 ***********************************************************************/
#include <inttypes.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <poll.h>
#include <stdarg.h>

#include <setjmp.h>

#include "tss2_mu.h"
#include "tss2_sys.h"
#include "tss2_tcti_device.h"

#include "tss2-tcti/tcti-common.h"
#include "tss2-tcti/tcti-device.h"

#define LOGMODULE fuzz
#include "tss2_tcti.h"
#include "util/log.h"
#include "test.h"
#include "test-options.h"
#include "context-util.h"
#include "tss2-sys/sysapi_util.h"
#include "tss2-tcti/tcti-fuzzing.h"

int
test_invoke (
        TSS2_SYS_CONTEXT *sysContext)
{
    TPM2_HANDLE objectHandle;
    TPM2B_NAME name;

    Tss2_Sys_LoadExternal_Complete (
        sysContext,
        &objectHandle,
        &name
    );

    return 0;
}
```

### Sample `_Prepare`

`fuzz_fill` `memcpy`'s fuzzer generated data into the memory of the structs being used as arguments.

```c
/* SPDX-License-Identifier: BSD-2 */
/***********************************************************************
 * Copyright (c) 2018, Intel Corporation
 *
 * All rights reserved.
 ***********************************************************************/
#include <inttypes.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <poll.h>
#include <stdarg.h>

#include <setjmp.h>

#include "tss2_mu.h"
#include "tss2_sys.h"
#include "tss2_tcti_device.h"

#include "tss2-tcti/tcti-common.h"
#include "tss2-tcti/tcti-device.h"

#define LOGMODULE fuzz
#include "tss2_tcti.h"
#include "util/log.h"
#include "test.h"
#include "test-options.h"
#include "context-util.h"
#include "tss2-sys/sysapi_util.h"
#include "tss2-tcti/tcti-fuzzing.h"

int
test_invoke (
        TSS2_SYS_CONTEXT *sysContext)
{
    int ret;
    const TPM2B_SENSITIVE inPrivate;
    const TPM2B_PUBLIC inPublic;
    TPMI_RH_HIERARCHY hierarchy;

    ret = fuzz_fill (
        sysContext,
        6,
        sizeof (inPrivate), &inPrivate,
        sizeof (inPublic), &inPublic,
        sizeof (hierarchy), &hierarchy
    );
    if (ret) {
        return ret;
    }

    Tss2_Sys_LoadExternal_Prepare (
        sysContext,
        &inPrivate,
        &inPublic,
        hierarchy
    );

    return EXIT_SUCCESS;
}
```

**Comment by flihp** [comment]
This is great. Spent some time going through your patches today. Will be doing this periodically over the next few. Initial feedback: you need to squash some of your commits. In reviewing the integration with autotools I've run across a few nits to pick only to find you cleaned them up in another commit down the queue 63c44533e99d0f19b88371e897794f8ef5d560eb. Reviewing patches that themselves change down the queue is very difficult (see `ENABLE_TCTI_FUZZING` conditional in Makefile.am).

File: test/fuzz/main-sapi.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        .socket_port    = PORT_DEFAULT,
+    };
+
+    get_test_opts_from_env (&opts);
+    if (sanity_check_test_opts (&opts) != 0) {
+        LOG_ERROR("Checking test options");
+        exit(1); /* fatal error */
+    }
+    /* Use the fuzzing tcti */
+    opts.tcti_type = FUZZING_TCTI;
```
Comment (pdxjohnny) [inline test/fuzz/main-sapi.cpp]: move this to opts invocation

File: script/fuzz-log-compiler.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+#;**********************************************************************;
+# Copyright (c) 2019, Intel Corporation
```
Comment (pdxjohnny) [inline script/fuzz-log-compiler.sh]: - [ ] Just use SPDX header

File: bootstrap [hunk 3]
```diff
@@ -63,4 +63,12 @@ echo "Generating file lists: ${VARS_FILE}"
   printf "TSS2_MU_SRC = \$(TSS2_MU_C) \$(TSS2_MU_H)"
 ) > ${VARS_FILE}
 
+# Generate fuzz tests unless environment variable NO_GEN_FUZZ is set to 1
```
Comment (AndreasFuchsSIT) [inline bootstrap]: I think, the default should be "fuzz-free", i.e. for maintainers and users.
This I'd like to see the default case reversed.

File: bootstrap [hunk 4]
```diff
@@ -63,4 +63,12 @@ echo "Generating file lists: ${VARS_FILE}"
...
+if test "${NO_GEN_FUZZ}0" -ne 10; then
```
Comment (AndreasFuchsSIT) [inline bootstrap]: "x${ABC}" -ne "x1" is way more common than appending a 0...

File: configure.ac [hunk 5]
```diff
@@ -10,7 +10,8 @@ AC_INIT([tpm2-tss],
         [https://github.com/tpm2-software/tpm2-tss])
 AC_CONFIG_MACRO_DIR([m4])
 : ${CFLAGS=""}
-AC_PROG_CC
+AC_PROG_CXX([clang++ g++])
```
Comment (AndreasFuchsSIT) [inline configure.ac]: Are we sure that we are not killing other compilers by this ?

File: configure.ac [hunk 6]
```diff
@@ -136,13 +137,26 @@ AC_ARG_ENABLE([tcti-device],
             [enable_tcti_device=$enableval],
             [enable_tcti_device=yes])
 AM_CONDITIONAL([ENABLE_TCTI_DEVICE], [test "x$enable_tcti_device" != xno])
+AM_COND_IF([ENABLE_TCTI_DEVICE], [AC_DEFINE([HAVE_TCTI_DEVICE], [1],
```
Comment (AndreasFuchsSIT) [inline configure.ac]: In all other places we are using
```
AS_IF([test "x$enable_tcti_device_async" = "xyes"],
AC_DEFINE([TCTI_ASYNC],[1]))
```
We should stay consistent.

File: configure.ac [hunk 7]
```diff
@@ -136,13 +137,26 @@ AC_ARG_ENABLE([tcti-device],
...
+           [Define if building tcti-device module.])], [])
 
 AC_ARG_ENABLE([tcti-mssim],
             [AS_HELP_STRING([--enable-tcti-mssim],
                             [build the tcti-mssim module (default is yes)])],
             [enable_tcti_mssim=$enableval],
             [enable_tcti_mssim=yes])
 AM_CONDITIONAL([ENABLE_TCTI_MSSIM], [test "x$enable_tcti_mssim" != xno])
+AM_COND_IF([ENABLE_TCTI_MSSIM], [AC_DEFINE([HAVE_TCTI_MSSIM], [1],
```
Comment (AndreasFuchsSIT) [inline configure.ac]: see above

File: configure.ac [hunk 8]
```diff
@@ -136,13 +137,26 @@ AC_ARG_ENABLE([tcti-device],
...
+           [Define if building tcti-mssim module.])], [])
+
+AC_ARG_ENABLE([tcti-fuzzing],
+            [AS_HELP_STRING([--enable-tcti-fuzzing],
+                            [build the tcti-fuzzing module (default is no)])],
+            [enable_tcti_fuzzing=$enableval],
+            [enable_tcti_fuzzing=no])
+AM_CONDITIONAL([ENABLE_TCTI_FUZZING], [test "x$enable_tcti_fuzzing" != xno])
+AM_COND_IF([ENABLE_TCTI_FUZZING], [AC_DEFINE([HAVE_TCTI_FUZZING], [1],
```
Comment (AndreasFuchsSIT) [inline configure.ac]: see above

File: script/gen_fuzz.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python3
```
Comment (AndreasFuchsSIT) [inline script/gen_fuzz.py]: This file is lacking a single line of comment.
Even a blank line between the hunge strings would be of help.

Maybe you could at least add a general description to the top of this.

File: src/tss2-sys/sysapi_util.h [hunk 10]
```diff
@@ -64,8 +64,6 @@ typedef struct {
     size_t nextData;
 } _TSS2_SYS_CONTEXT_BLOB;
 
-struct TSS2_SYS_CONTEXT;
```
Comment (AndreasFuchsSIT) [inline src/tss2-sys/sysapi_util.h]: This looks like a random change that sould be its own commit (sorry for the pickiness)

File: test/fuzz/tcti/tss2-tcti-fuzzing.map [hunk 11]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (AndreasFuchsSIT) [inline test/fuzz/tcti/tss2-tcti-fuzzing.map]: Do we need a map file for a noinst library ?
This you can just export all symbols, who cares ?

File: test/fuzz/tcti/tss2_tcti_fuzzing.h [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+#ifndef TSS2_TCTI_FUZZING_H
+#define TSS2_TCTI_FUZZING_H
+
+#include "tss2_tcti.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+TSS2_RC tcti_platform_command(
```
Comment (AndreasFuchsSIT) [inline test/fuzz/tcti/tss2_tcti_fuzzing.h]: Is this function actually called an/or used anywhere ?
tcti_device also does not have a platform_command function. That one should be special for tcti_mssim only.

File: bootstrap (hunk 3 above)
Comment (pdxjohnny) [reply bootstrap]: Will do

File: .travis.yml [hunk 13]
```diff
@@ -117,6 +117,15 @@ script:
     else
       make -j$(nproc) check
     fi
+# check fuzz targets
```
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply .travis.yml]: - This only runs them for a single iteration so make sure they compiled, linked, and run correctly
- I'll be submitting this to OSS Fuzz so that it does get run 24/7
- [ ] John will check out crypto backends to see how that works

File: test/fuzz/tcti/tss2-tcti-fuzzing.pc.in [hunk 14]
```diff
@@ -0,0 +1,7 @@
+Name: tss2-tcti-fuzzing
```
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply test/fuzz/tcti/tss2-tcti-fuzzing.pc.in]: Sounds good, wasn't sure when I did this so I added it anyway, will change

File: Dockerfile [hunk 15]
```diff
@@ -46,11 +46,29 @@ RUN apt-get install -y \
...
+    python3 \
+    clang
 
-# TPM2-TSS
+# TPM Fuzzing
 COPY . /tmp/tpm2-tss/
 WORKDIR /tmp/tpm2-tss
+
+## Fuzzing
+FROM base AS fuzzing
```
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply Dockerfile]: Yes this keeps the existing behavior of the "regular" build

File: configure.ac (hunk 5 above)
Comment (pdxjohnny) [reply configure.ac]: My understanding is that this is an order of preference search list, aka it will try to grab clang before gcc if both are present on a system.

File: bootstrap (hunk 4 above)
Comment (pdxjohnny) [reply bootstrap]: That doesn't seem to work, I think because we are in sh instead of bash. Okay if i leave this way?

File: src/tss2-sys/sysapi_util.h (hunk 10 above)
Comment (pdxjohnny) [reply src/tss2-sys/sysapi_util.h]: See #1244

File: Dockerfile (hunk 15 above)
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply Dockerfile]: It does build the fuzzing targets too, but the tagged image is the same as it used to be. The `AS` allows someone to build fuzzing and not unit as well as keep the default of just unit

File: .travis.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply .travis.yml]: thanks for clarifying

File: Dockerfile (hunk 15 above)
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply Dockerfile]: ok, thanks for the info

File: configure.ac (hunk 5 above)
Comment (AndreasFuchsSIT) [reply configure.ac]: Understood. My question is about CCs that are neither clang or gcc. Will they be precluded this way ?

Comment (pdxjohnny) [reply configure.ac]: I don't think so. But I've gone ahead and reverted to the pervious way of no args because then CC must be given to configure which is better I think because it preserves original behavior and is explicit.

File: script/gen_fuzz.py (hunk 9 above)
Comment (pdxjohnny) [reply script/gen_fuzz.py]: I have added comments

**Comment by pdxjohnny** [comment]
@AndreasFuchsSIT @flihp @tstruk I have made the changes, including making separate commits. I force push'd. I don't know if there is a better way, let me know if you know for future.

**Comment by pdxjohnny** [comment]
@AndreasFuchsSIT When you get a chance I've updated this to work with the changes that happened in master

**Comment by pdxjohnny** [comment]
Hi @AndreasFuchsSIT sorry for pinging you a bunch, but I'd heard you had been on vacation so I thought I'd give you a ping again now that you're back. This is ready for a re-review. I made all the changes you requested and separated into commits.

**Comment by AndreasFuchsSIT** [comment]
So programming-wise it's much better. Documentation would be nice of course. So no objections from my side...

Content-wise, I still have no experience at all...

**Comment by pdxjohnny** [comment]
Thanks!

File: test/fuzz/tcti/tcti-fuzzing.c [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+    data = tcti_fuzzing->data;
+    size = tcti_fuzzing->size;
+
+    va_start (ap, count);
+
+    for (i = 0U; i < (count / 2); ++i) {
+        length = va_arg (ap, size_t);
+        dest = va_arg (ap, void *);
+        if (size > (combined + length)) {
+            curr = &data[combined];
```
Comment (tstruk) [inline test/fuzz/tcti/tcti-fuzzing.c]: I think that be just:
curr = data;

va_arg() performs the iteration. If one would do something like:

TPMI_DH_OBJECT A = {0};
TPMI_DH_OBJECT B = {0};

fuzz_fill (sysContext, 4, sizeof (B), &B, sizeof(A), &A));

then the &data[combined] in the second iteration would point somewhere after the object B.

File: .travis.yml [hunk 17]
```diff
@@ -106,9 +106,9 @@ script:
...
+      scan-build ../configure --enable-tcti-partial-reads=$WITH_TCTI_PARTIAL --enable-tcti-device-async=$WITH_TCTI_ASYNC --enable-tcti-fuzzing --enable-unit --enable-integration --with-crypto=$WITH_CRYPTO $CONFIGURE_OPTIONS CFLAGS=-I${PWD}/../osslinstall/usr/local/include LDFLAGS=-L${PWD}/../osslinstall/usr/local/lib
     else
-      ../configure --enable-tcti-partial-reads=$WITH_TCTI_PARTIAL --enable-tcti-device-async=$WITH_TCTI_ASYNC --enable-unit --enable-integration --with-crypto=$WITH_CRYPTO $CONFIGURE_OPTIONS CFLAGS=-I${PWD}/../osslinstall/usr/local/include LDFLAGS=-L${PWD}/../osslinstall/usr/local/lib
+      ../configure --enable-tcti-partial-reads=$WITH_TCTI_PARTIAL --enable-tcti-device-async=$WITH_TCTI_ASYNC --enable-tcti-fuzzing --enable-unit --enable-integration --with-crypto=$WITH_CRYPTO $CONFIGURE_OPTIONS CFLAGS=-I${PWD}/../osslinstall/usr/local/include LDFLAGS=-L${PWD}/../osslinstall/usr/local/lib
```
Comment (tstruk) [inline .travis.yml]: Isn't that what the commit 5 "travis: Build and check fuzz tests" does? this commit seams to be redundant.

File: configure.ac [hunk 18]
```diff
@@ -148,6 +150,15 @@ AM_CONDITIONAL([ENABLE_TCTI_MSSIM], [test "x$enable_tcti_mssim" != xno])
 AS_IF([test "x$enable_tcti_mssim" = "xyes"],
 	AC_DEFINE([TCTI_MSSIM],[1], [TCTI FOR MS SIMULATOR]))
 
+AC_ARG_ENABLE([tcti-fuzzing],
```
Comment (tstruk) [inline configure.ac]: Can the tcti-fuzzing coexist with other tctis? Don't we need to make sure that other tctis are disabled when fuzzing is enabled?

File: test/fuzz/tcti/tcti-fuzzing.c (hunk 16 above)
Comment (pdxjohnny) [reply test/fuzz/tcti/tcti-fuzzing.c]: Note after in person review: John needs to change the names here

File: .travis.yml (hunk 17 above)
Comment (pdxjohnny) [reply .travis.yml]: AR for John remove 2nd commit

File: configure.ac (hunk 18 above)
Comment (pdxjohnny) [reply configure.ac]: AR for John add checks to configure to make sure if TCTI Fuzzing is built, it is the only TCTI being built.

File: configure.ac [hunk 19]
```diff
@@ -327,14 +362,23 @@ AC_OUTPUT
 
 AM_COND_IF([ENABLE_TCTI_DEVICE], [],
            [AM_COND_IF([ENABLE_TCTI_MSSIM], [],
-                       [AC_MSG_WARN("No build-in TCTI module enabled")])])
+                       [AM_COND_IF([ENABLE_TCTI_FUZZING], [],
+                                   [AC_MSG_WARN("No build-in TCTI module enabled")])])])
+
+AM_COND_IF([ENABLE_TCTI_FUZZING], [
+            AM_COND_IF([ENABLE_TCTI_DEVICE],
+                       AC_MSG_ERROR("Fuzzing TCTI is meant to be built as the only TCTI"), [])
```
Comment (pdxjohnny) [inline configure.ac]: @tstruk I made those three changes, remove 2nd commit, change variable names, and add guards to configure.ac to make sure fuzzing TCTI isn't used with either of the others.

**Comment by pdxjohnny** [comment]
Thank you!!
